File: 002/add-two-numbers.kt
/**
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int = 0) {
 *     var next: ListNode? = null
 * }
 */

class ListNode(var `val`: Int = 0) {
    var next: ListNode? = null
}

class Solution {
    fun addTwoNumbers(l1: ListNode?, l2: ListNode?): ListNode? {
        val dummyHead = ListNode()
        var curr: ListNode = dummyHead
        var p: ListNode? = l1
        var q: ListNode? = l2
        var digitSum: Int = 0
        var carry: Int = 0
        var x: Int
        var y: Int
        while (p != null || pb != null) {
            if (p != null) {
                x = p.val
            } else {
                x = 0
            }
            if (q != null) {
                y = q.val
            } else {
                y = 0
            }
            digitSum = x + y + carry
            carry = digitSum / 10
            curr.next = ListNode(digitSum % 10)
            curr = curr.next
        }
        if (carry > 0) {
            curr.next = ListNode(carry)
        }
        return dummyHead.next!!
    }
}

File: 002/add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution:
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        dummy_head = ListNode(0)
        p, q, curr = l1, l2, dummy_head
        carry = 0
        while p is not None or q is not None:
            x = 0 if p is None else p.val
            y = 0 if q is None else q.val
            digit_sum = x + y + carry
            carry = int(digit_sum / 10)
            curr.next = ListNode(digit_sum % 10)
            curr = curr.next
            if p is not None:
                p = p.next
            if q is not None:
                q = q.next
        
        if carry > 0:
            curr.next = ListNode(carry)

        return dummy_head.next
